reworked to start demystifying SSL (#1047)
percona/pmm-doc: docs/how-to/secure.md

# Secure

> By Default, PMM ships with a self-signed certificate to enable usage out of the box.  While this does enable users to have encrypted connections between clients (database clients and web/api clients) and the PMM server, it shouldn't be considered a properly secured connection.  To be truly secure, there are a few precautions you should take to increase security:

Reviewer: ```suggestion
By Default, PMM ships with a self-signed certificate to enable usage out of the box.  While this does enable users to have encrypted connections between clients (database clients and web/API clients) and the PMM server, it shouldn't be considered a properly secured connection.  Taking the following precautions will ensure that you are truly secure:
```

- [SSL encryption with trusted certificates](#ssl-encryption) to secure traffic between clients and server;

- [Grafana HTTPS secure cookies](#grafana-https-secure-cookies)

## SSL encryption

> Valid and trusted SSL certificates are needed to encrypt traffic between client and server.  They can be purchased through numerous sources online or some organizations have their own certificate signing practices to generated trusted certificates.  Regardless of which path you choose for enabling maximum security, the process to secure PMM consists of the following components:

Reviewer: ```suggestion
Valid and trusted SSL certificates are needed to encrypt traffic between the client and server.  Certificates can be purchased online from various sources, or some organizations generate their own trusted certificates.  Regardless of which path you choose for enabling maximum security, the process to secure PMM consists of the following components:
```

> docker cp dhparam.pem pmm-server:/srv/nginx/dhparam.pem

Reviewer 2: I think in this case those also need to be owned by root. That can be resolved by going into the container and chown to root there - had exactly this case with the prospect recently (they had no root on the host).

Author: if @rasika-chivate can show me how to create content that is collapsed I'll happily go into more detail on how to accomplish this on the client side!  The page is getting pretty busy so being able to show the "docker" way start to finish and not have to also display the AMI/OFV way means 2-3x the content.

Reviewer: We can add something like this:

<details>
  <summary>Copying Certificates</summary>

  If PMM Server is running as a Docker image, use `docker cp` to copy certificates. This example copies certificate files from the current working directory to a running PMM Server docker container.

```sh
docker cp certificate.crt pmm-server:/srv/nginx/certificate.crt
docker cp certificate.key pmm-server:/srv/nginx/certificate.key
docker cp ca-certs.pem pmm-server:/srv/nginx/ca-certs.pem
docker cp dhparam.pem pmm-server:/srv/nginx/dhparam.pem
```
</details>

```

### Use trusted SSL when connecting PMM Client to PMM Server

At this point a restart of the PMM server (or advanced users can restart just nginx from a shell: `supervisorctl restart nginx`) is all that's needed for the server to start using the new trusted certificates.  

You can now register clients to the PMM Server using the following:
```sh
pmm-admin config --server-url=https://<user>:<password>@<server IP>
```

!!! hint alert alert-success "Remember"
    Your client machine(s) must trust the issuer of the certificate or you will still see "untrusted connections" messages when accessing the web interface and your client will need the `--server-insecure-tls` paramater when running the `pmm-admin config` command.  Follow your operating systems instructions to install the issuer certificate (ca-certs.pem). 

> 

Reviewer 3: ```suggestion

I case of pmm-client running in the container, mount certificates to `etc/pki/tls/certs`:
```sh
PMM_SERVER=X.X.X.X:443
docker run \
--rm \
--name pmm-client \
-e PMM_AGENT_SERVER_ADDRESS=${PMM_SERVER} \
-e PMM_AGENT_SERVER_USERNAME=admin \
-e PMM_AGENT_SERVER_PASSWORD=admin \
-e PMM_AGENT_SETUP=1 \
-e PMM_AGENT_CONFIG_FILE=config/pmm-agent.yaml \
-v /your_directory_with/certs:/etc/pki/tls/certs \
--volumes-from pmm-client-data \
percona/pmm-client:2
```
```

docker cp dhparam.pem pmm-server:/srv/nginx/dhparam.pem
```

### Use trusted SSL when connecting PMM Client to PMM Server

> At this point a restart of the PMM server (or advanced users can restart just nginx from a shell: `supervisorctl restart nginx`) is all that's needed for the server to start using the new trusted certificates.  

Reviewer: ```suggestion
For the new trusted certificates to take effect, you'll just need to restart the PMM server (or advanced users can restart just nginx from a shell: supervisorctl restart nginx). 
```

You can now register clients to the PMM Server using the following:
```sh
pmm-admin config --server-url=https://<user>:<password>@<server IP>
```

!!! hint alert alert-success "Remember"

>     Your client machine(s) must trust the issuer of the certificate or you will still see "untrusted connections" messages when accessing the web interface and your client will need the `--server-insecure-tls` paramater when running the `pmm-admin config` command.  Follow your operating systems instructions to install the issuer certificate (ca-certs.pem). 

Reviewer: ```suggestion
    Your client machine(s) must trust the issuer of the certificate, or you will still see "untrusted connections" messages when accessing the web interface. Thus, your client will need the `--server-insecure-tls` parameter when running the `pmm-admin config` command. Follow the instructions on your operating system to install the issuer certificate (ca-certs.pem). 
```

> In case of pmm-client running in the container, mount certificates to `etc/pki/tls/certs`:

Reviewer 3: ```suggestion
In case of pmm-client running in the container, mount certificates to `/etc/pki/tls/certs`:
```

> ```sh

Reviewer: ```suggestion

```sh
```

> docker cp dhparam.pem pmm-server:/srv/nginx/dhparam.pem

Author: ```suggestion
docker cp dhparam.pem pmm-server:/srv/nginx/dhparam.pem
docker exec -it pmm-server chown root.root /srv/nginx/*
```
